### ISSUE TITLE
Json Unmarshaller ignores error around reading 'data' when 'datacontenttype' is invalid

### ISSUE THREAD

**Issue by agriess** [cloudevents/sdk-go#782]
Using: github.com/cloudevents/sdk-go/v2 v2.10.1

This issue seems to be affected by the ordering of cloud event attributes in a json http request. When the 'data' attribute is before the  invalid 'datacontenttype' attribute, an error is not returned and the resulting cloud event is missing its data. This can be observed in the following unit test

	import (
		"github.com/cloudevents/sdk-go/v2/event"
		"github.com/stretchr/testify/assert"
		"testing"
	)
	
	func TestCloudEventUnmarshalling(t *testing.T) {
	
		structuredEventWithDataAfterDataContentType := `{
				"specversion": "1.0",
				"datacontenttype": "application",
				"id": "test123",
				"source": "/test/source",
				"type": "test.event.type",
				"time": "2021-07-21T01:00:59.365Z",
				"data": {
					"test": "test"
				}
			}`
	
		testEvent1 := &event.Event{}
		err := testEvent1.UnmarshalJSON([]byte(structuredEventWithDataAfterDataContentType))
		assert.NotNil(t, err)
	
		structuredEventWithDataBeforeDataContentType := `{
				"specversion": "1.0",
				"data": {
					"test": "test"
				},
				"datacontenttype": "application",
				"id": "test123",
				"source": "/test/source",
				"type": "test.event.type",
				"time": "2021-07-21T01:00:59.365Z"
			}`
	
		testEvent2 := &event.Event{}
		err = testEvent2.UnmarshalJSON([]byte(structuredEventWithDataBeforeDataContentType))
		assert.Nil(t, err)
		assert.Empty(t, testEvent2.Data())
	
	}

In the case where an error message is returned, it doesn't seem correct either. The original error message: 
"ReadString: expects " or n, but found {, error found in #10 byte of ...|	"data": {
					"tes|..., bigger context ...|	"time": "2021-07-21T01:00:59.365Z",
				"data": {
					"test": "test"
				}
			}|..."

which is what I expect gets overridden with:
"expect { or , or } or n, but found ", error found in #10 byte of ...|: {
					"test": "te|..., bigger context ...|: "2021-07-21T01:00:59.365Z",
				"data": {
					"test": "test"
				}
			}|..."

**Comment by n3wscott** [comment]
I would expect this error: 

```
[decode] unsupported content type: "application"
```

from https://github.com/cloudevents/sdk-go/blob/main/v2/event/datacodec/codec.go#L63-L68

So likely, there is no data in the bytes between 1 and 2 in the demo above, but I would expect it to throw an error. 

Could you look into why you are not getting the error on https://github.com/cloudevents/sdk-go/blob/main/v2/event/datacodec/codec.go#L77, there is no codec registered for `application`.

**Comment by n3wscott** [comment]
Interesting, I will investigate. Thanks for the issue report!

**Comment by n3wscott** [comment]
Thanks for the report. There was a case where the json iterator error was not being looked at for cases when we see data before content type. In this case we created a new json iterator to process the bytes. 

As for the error, I have added more context to the new error `Unmarshal` returns. And I was incorrect, the data codec does not get a chance here yet, this is pre-processing to convert a structured cloudevent into an `Event` object, and processing of the data field does not happen until you attempt to access the data via a `DataAs` method, or sending this event back out on a wire via the Bindings implementations.

We do this to speed things up, for example: if you had an xml string, we would not parse the data string into an xml document if the data is never accessed, we would leave it as a string and then sending it along again should be a quick matter (for middleware case); or we access the data using `event.DataAs` and the codecs become involved.
